docs/guide/checkpoints.md

```diff
@@ -187,29 +187,95 @@ for ds in dc.datasets():
     print(ds.name)
 ```
 
-## Limitations
+## UDF-Level Checkpoints
 
-- **Script-based:** Code must be run as a script (not interactively or as a module).
-- **Hash-based matching:** Any change to the chain will create a different hash, preventing checkpoint reuse.
-- **Same script path:** The script must be run from the same absolute path for parent job linking to work.
+DataChain automatically creates checkpoints for UDF operations (`.map()`, `.gen()`, `.agg()`), not just at `.save()` calls. For `.map()` and `.gen()` operations, **DataChain saves processed rows continuously during UDF execution**, not only when the UDF completes. If your script fails partway through a UDF operation, the next run will skip already-processed rows and continue where it left off - even if you've modified the UDF code to fix a bug.
 
-## Future Plans
+**Note:** For `.agg()` operations, checkpoints are created when the aggregation completes successfully, but partial results are not tracked. If an aggregation fails partway through, it will restart from scratch on the next run.
 
-### UDF-Level Checkpoints
+### How It Works
 
-Currently, checkpoints are created only when datasets are saved using `.save()`. This means that if a script fails during a long-running UDF operation (like `.map()`, `.gen()`, or `.agg()`), the entire UDF computation must be rerun on the next execution.
+When executing `.map()` or `.gen()` operations, DataChain:
 
-Future versions will support **UDF-level checkpoints**, creating checkpoints after each UDF step in the chain. This will provide much more granular recovery:
+1. **Saves processed rows incrementally** as the UDF processes your dataset
+2. **Creates a checkpoint** when the UDF operation completes successfully
+3. **Allows you to fix bugs and continue** - if the UDF fails, you can modify the code and re-run, skipping already-processed rows
+4. **Invalidates the checkpoint if you change the UDF after successful completion** - completed UDFs are recomputed from scratch if the code changes
+
+For `.agg()` operations, checkpoints are only created upon successful completion, without incremental progress tracking.
+
+### Example: Fixing a Bug Mid-Execution
 
 ```python
-# Future behavior with UDF-level checkpoints
+def process_image(file):
+    # Bug: this will fail on some images
+    img = Image.open(file.get_local_path())
+    return {"width": img.size[0], "height": img.size[1]}
+
 result = (
-    dc.read_csv("data.csv")
-    .map(heavy_computation_1)  # Checkpoint created after this UDF
-    .map(heavy_computation_2)  # Checkpoint created after this UDF
-    .map(heavy_computation_3)  # Checkpoint created after this UDF
-    .save("result")
+    dc.read_dataset("images")
+    .map(process_image, output={"width": int, "height": int})
```

Comment:
simplify datachain code here, add types to UDF, use tuples, etc - make it simpler

check other places also